### PR TITLE
routedns: init at 0.1.5

### PR DESCRIPTION
**Pull request by jsimonetti**
###### Description of changes

Add routedns package and module. Also add myself as maintainer.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -14635,4 +14635,11 @@
     github = "snpschaaf";
     githubId = 105843013;
   };
+  jsimonetti = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:14638]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: nixos/modules/services/networking/routedns.nix [hunk 2]
```diff
@@ -0,0 +1,82 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+with lib; let
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/routedns.nix:7]:
```suggestion
{ config
, lib
, pkgs
, ...
}:

with lib;

let
```

File: nixos/modules/services/networking/routedns.nix [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+          };
+          listeners.local-udp = {
+            address = "127.0.0.1:53";
+                   protocol = "udp";
+                   resolver = "cloudflare-cached";
+          };
+          listeners.local-tcp = {
+            address = "127.0.0.1:53";
+                   protocol = "tcp";
+                   resolver = "cloudflare-cached";
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/routedns.nix:34]:
```suggestion
            protocol = "udp";
            resolver = "cloudflare-cached";
          };
          listeners.local-tcp = {
            address = "127.0.0.1:53";
            protocol = "tcp";
            resolver = "cloudflare-cached";
```

File: nixos/modules/services/networking/routedns.nix [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  config = mkIf cfg.enable {
+    # Prefer the config file over settings if both are set.
+    services.routedns.configFile = mkDefault (settingsFormat.generate "routedns.toml" cfg.settings);
+
+    systemd.services.routedns = {
+      description = "RouteDNS - DNS stub resolver, proxy and router";
+      after = ["network.target"]; # in case a bootstrap resolver is used, this might fail a few times until the respective server is actually reachable
+      wantedBy = ["multi-user.target"];
+      wants = ["network-online.target"];
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/routedns.nix:66]:
```suggestion
      wants = ["network.target"];
```
network-online does not mean we have internet connection

File: pkgs/tools/networking/routedns/default.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{
+  lib,
+  buildGoModule,
+  fetchFromGitHub,
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/routedns/default.nix:4]:
```suggestion
{ lib
, buildGoModule
, fetchFromGitHub
```

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -35388,4 +35388,8 @@ with pkgs;
   honeyvent = callPackage ../servers/tracing/honeycomb/honeyvent { };
 
   mictray = callPackage ../tools/audio/mictray { };
+
+  routedns = callPackage ../tools/networking/routedns {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:35392]: Please sort this into a fitting category and sort there alphabetically.

File: pkgs/tools/networking/routedns/default.nix [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+}:
+buildGoModule rec {
+  pname = "routedns";
+  version = "0.1.5";
+
+  src = fetchFromGitHub {
+    owner = "folbricht";
+    repo = "routedns";
+    rev = "02f14a567fee2a289810979446f5260b8a31bf73";
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/routedns/default.nix:13]:
```suggestion
    # https://github.com/folbricht/routedns/issues/237
    rev = "02f14a567fee2a289810979446f5260b8a31bf73";
```

File: pkgs/tools/networking/routedns/default.nix [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    sha256 = "sha256-oImimNBz1qizUPD6qHi73fGKNCu5cii99GIUo21e+bs=";
+  };
+
+  vendorSha256 = "sha256-T6adpxJgOPGy+UOOlGAAf1gjk1wJxwOc9enfv9X3LBE=";
+
+  subPackages = ["./cmd/routedns"];
+
+  meta = with lib; {
+    homepage = "https://github.com/folbricht/routedns";
+    description = "RouteDNS - DNS stub resolver, proxy and router";
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/routedns/default.nix:23]:
```suggestion
    description = "DNS stub resolver, proxy and router";
```

**Comment by jsimonetti** [comment]
@SuperSandro2000 I did my best to address your comments. Please let me know if I need to make additional changes. Thank you!

File: pkgs/tools/networking/routedns/default.nix [hunk 9]
```diff
@@ -0,0 +1,29 @@
+{ lib
+, buildGoModule
+, fetchFromGitHub
+,
+}:
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/routedns/default.nix:5]:
```suggestion
}:

```

**Comment by SuperSandro2000** [comment]
Package LGTM, I haven't looked at the module.

**Comment by jsimonetti** [comment]
If we could backport this to 22.05 that would be great.

File: nixos/modules/services/networking/routedns.nix [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      default = pkgs.routedns;
+      defaultText = literalExpression "pkgs.routedns";
+      type = types.package;
+      description = "RouteDNS package to use.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    # Prefer the config file over settings if both are set.
+    services.routedns.configFile = mkDefault (settingsFormat.generate "routedns.toml" cfg.settings);
```
Comment (aanderse) [inline nixos/modules/services/networking/routedns.nix:63]: Put this in `configFile.default` instead. Simpler because the module system takes care of it for you, no need for `mkDefault`.

File: nixos/modules/services/networking/routedns.nix [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+    systemd.services.routedns = {
```
Comment (aanderse) [inline nixos/modules/services/networking/routedns.nix:65]: Your `systemd` service seems superior to the service provided by [upstream](https://github.com/folbricht/routedns/blob/master/cmd/routedns/routedns.service). Please consider opening a PR upstream to improve it.

**Review by aanderse — APPROVED** [review]
Nice simple module :+1:

File: nixos/modules/services/networking/routedns.nix (hunk 10 above)
Comment (jsimonetti) [reply nixos/modules/services/networking/routedns.nix:63]: I'm not familiar with that way. Do you mean this?
`services.routedns.configFile.default =  settingsFormat.generate "routedns.toml" cfg.settings;`

or should I set `options.services.routedns.configFile.default =  settingsFormat.generate "routedns.toml" cfg.settings;` instead?

Comment (aanderse) [reply nixos/modules/services/networking/routedns.nix:63]:
```
options = {
  services.routedns = {
    configFile = mkOption {
      default = settingsFormat.generate "routedns.toml" cfg.settings;
      defaultText = "A routedns configuration file automatically generated by values from services.routedns.*";
    };
  };
};
```

**Comment by SuperSandro2000** [comment]
Please squash the review commits into the right init commit.

**Comment by jsimonetti** [comment]
Is this ready for merge?
Please let me know if I am forgetting something. Thanks!
